api/maven-api-model/src/main/mdo/maven.mdo

```diff
@@ -2646,6 +2646,15 @@
             ]]>
           </description>
         </field>
+        <field>
+          <name>priority</name>
+          <version>4.2.0+</version>
```

Comment:
IIUC, this is a new XML element, then not compatible with previous pom.xml schema
is this change for Maven 4 only? and with proper removing in consumer POM?

did you try to define the field as XML attribute to avoid the issue (and be able to add the feature to Maven 3.9)?

Reply:
Not at all.

This element can not be removed if present in a POM, as this will change the semantics of the POM.  The build/consumer POM is all about inferred (or removal of unused) data.  If we were to upload a project which defines such a new element/attribute but use it with a maven version that does not know about it, the build results could be different and that would be bad.  So I don't think we should upload a POM stripped of this element.

So I think a new POM is needed for that one. 4.1.0 is kinda used internally for the api v4 migration, so I went with 4.2.0.  This field is a bit specific as it does already exist in the 4.0.0 model, but it is flagged as `xml.transient=true`, which means it is not read from the xml, but there's a field + getter/setter already generated.

On the compatibility side, I think this defines a new 4.2.0 POM.  Note that the way the xml reader is generated, it has two modes: the first mode when reading projects being built where a strict validation is used, i.e. any unknown element will be rejected and cause a parsing failure.  When using this mode, i.e. when building a project with a 4.2.0 `modelVersion`, maven 3.x will reject it.  However, if uploaded on maven central and used as a dependency, the lenient parsing mode is used, where any unknown element or attribute is simply ignored (see the `checkUnknownElement` and `checkUnknownAttribute` of the `MavenXpp3Reader`).  That will mean that projects with a modelVersion of 4.2.0 will be able to be consumed by older maven 3.x tools / projects.

Reply:
Maven is not the only one using these pom.xml, but many tools in the wild: that's why we always avoided adding XML elements, but only added XML attributes when really necessary = the only accepted changes (in addition to the "import" value for dependencyManagement scope
see attributes in https://maven.apache.org/ref/3.9.2/maven-model/maven.html

on build vs consumer, we'll need to again define what is consumer, but to me, this strict convention on available elements in published POMs will remain forever

Reply:
> Maven is not the only one using these pom.xml, but many tools in the wild: that's why we always avoided adding XML elements, but only added XML attributes when really necessary = the only accepted changes (in addition to the "import" value for dependencyManagement scope see attributes in https://maven.apache.org/ref/3.9.2/maven-model/maven.html
> 
> on build vs consumer, we'll need to again define what is consumer, but to me, this strict convention on available elements in published POMs will remain forever

What tools to you have in mind ? If our own parser can be quite lenient, I don't see why other tools would not be able to do the same.  We can't just decide to never update the model just because _other tools_ are using it.

Reply:
And what's the difference between adding an attribute and adding an element ? In our own parser, both are treated the same way (either discarded or rejected, depending if it's in strict/lenient mode).

Reply:
it is the convention that has been defined from the beginning for the whole consumer ecosystem, and the reason build bom vs consumer pom has been introduced https://cwiki.apache.org/confluence/display/MAVEN/Build+vs+Consumer+POM

we can test options and go back to the M dev ML to have feedback from others: this change is not only about our internal code

Reply:
> it is the convention that has been defined from the beginning for the whole consumer ecosystem, and the reason build bom vs consumer pom has been introduced https://cwiki.apache.org/confluence/display/MAVEN/Build+vs+Consumer+POM
> 
> we can test options and go back to the M dev ML to have feedback from others: this change is not only about our internal code

I think the initial assumption and driver for the build/consumer feature, i.e. that "changing the POM version and publishing artifacts on Maven Central with this new model would break consumers using either older Maven versions or other build tools" was done with a wrong assumption.  I'll double check with more tests, but afaik, consumers using old versions of maven won't be broken when _using_ the artifacts as a dependency (they would be broken if using such a project as a parent though).  So yes, this definitely would mean that if you use a 4.2.0 modelVersion (though you should not be forced to unless you actually need something in it) imposes to use a recent maven when building a project that inherit it.

I'm not sure about other tools.

Anyway, I'll send a mail to dev@ to discuss that.